designcompose/src/main/java/com/android/designcompose/squoosh/SquooshRoot.kt

```diff
@@ -478,6 +417,110 @@ fun SquooshRoot(
+                // are needed.
+                if (transitionRoot != null) {
+                    // Ensure that the transition target also has the incoming width/height applied.
+                    transitionRoot.applyLayoutConstraints(constraints, density)
+
+                    // Layout the transition root tree.
+                    layoutTree(
+                        transitionRoot,
+                        rootLayoutId,
+                        layoutIdAllocator,
+                        layoutCache,
+                        layoutValueCache
+                    )
+
+                    // Record all of the animations in progress or to be started in `nextAnimations`.
+                    val nextAnimations = HashMap<Int, SquooshAnimationRenderingInfo>()
+
+                    // Process all of the animated actions from the InteractionState. These are generated by
+                    // performing actions that were specified in the design doc.
+                    for (animatedAction in animatedActions) {
+                        createOrUpdateAnimation(
+                            root, transitionRoot,
+                            interactionState, variantTransitions,
+                            currentAnimations, nextAnimations, animationValues,
+                            action = animatedAction
+                        )
+                    }
+
+                    // Now do the same for the variant animations. These are generated by enums specified by
+                    // the application state changing over time.
+                    for (variantTransition in variantTransitions.transitions.values.toList()) {
+                        createOrUpdateAnimation(
+                            root, transitionRoot,
+                            interactionState, variantTransitions,
+                            currentAnimations, nextAnimations, animationValues,
+                            variant = variantTransition
+                        )
+                    }
+
+                    // Evolve the animations list; it would be better to just remove everything that doesn't
+                    // have a key in transitions...
+                    currentAnimations.clear()
+                    currentAnimations.putAll(nextAnimations)
+
+                    Log.d(
+                        TAG,
+                        "Updating animations resulted in ${nextAnimations.size} animations"
+                    )
+                }
+
+                val currentAnimationJob = animationJob.animationJob
+                val needsAnimationJob = interactionState.animations.isNotEmpty() || variantTransitions.transitions.isNotEmpty()
+                val hasAnimationJob = currentAnimationJob != null && currentAnimationJob.isActive
+                if (needsAnimationJob && !hasAnimationJob) {
+                    animationJob.animationJob = interactionScope.launch(start = CoroutineStart.UNDISPATCHED) {
```

Comment:
I think I should be able to remove the "animation <-> compose" glue out of this file more easily now -- before it was a bunch of logic that had to be in a `LaunchedEffect`, but now it's more easily movable because it just needs a `CoroutineScope`.

In general in Squoosh, I am seeing that there's a bunch of code that doesn't care about Compose at all (it's just resolving variants, building trees, evaluating layout, computing animated transitions, etc) and then some glue code that uses all the Compose APIs. I'm hoping to enforce that separation more strictly, and then move the generic code to Rust.